pkg/intoto/metadata.go

```diff
@@ -9,89 +9,148 @@ import (
+		}
+
+		data, err := readFile(filename)
 		if err != nil {
-			return err
+			return nil, err
 		}
+
+		metadata := Metadata{Data: data}
+		custom := Custom{InToto: metadata}
+		filename = gun + "/in-toto-pubkeys/" + filename
+		publicKeys[filename] = custom
 	}
 
-	return nil
+	return publicKeys, nil
 }
 
-// GetMetadataRawMessage takes In-Toto metadata and returns a canonical RawMessage
-// that can be stored in the TUF targets custom field.
-//
-// TODO: layout signing key should not be passed by the library.
-// Layouts should be signed with the targets key used to sign the TUF collection.
-func GetMetadataRawMessage(layout string, linkDir string, layoutKey string) (canonicaljson.RawMessage, error) {
-	k, err := ioutil.ReadFile(layoutKey)
-	if err != nil {
-		return nil, fmt.Errorf("cannot get canonical JSON from file %v: %v", layoutKey, err)
+// GetRootLayout reads root layout off disk into a RootLayout struct
+func GetRootLayout(gun string, filename string, publicKeys PublicKeys) (RootLayout, error) {
+	var rootLayout RootLayout
+
+	if !strings.HasSuffix(filename, ".layout") {
+		return rootLayout, fmt.Errorf("%s does not have a .layout suffix", filename)
 	}
 
-	l, err := ioutil.ReadFile(layout)
+	data, err := readFile(filename)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get canonical JSON from file %v: %v", layout, err)
+		return rootLayout, err
+	}
+
+	filenames := make([]string, len(publicKeys))
+	for filename := range publicKeys {
+		filenames = append(filenames, filename)
 	}
 
-	links := make(map[string][]byte)
-	files, err := ioutil.ReadDir(linkDir)
+	metadata := Metadata{Data: data, PublicKeys: filenames}
+	custom := Custom{InToto: metadata}
+	filename = gun + "/in-toto-metadata/" + filename
+	rootLayout.Filename = filename
+	rootLayout.Custom = custom
+	return rootLayout, nil
+}
+
+// GetLinks reads link metadata off disk into a Links struct
+func GetLinks(gun string, dir string) (Links, error) {
+	fileinfos, err := ioutil.ReadDir(dir)
 	if err != nil {
-		return nil, fmt.Errorf("cannot read links directory %v: %v", linkDir, err)
+		return nil, fmt.Errorf("cannot read links directory %v: %v", dir, err)
 	}
-	for _, f := range files {
-		// TODO - Radu M
-		//
-		// robust check if file is actually a link
-		if !strings.Contains(f.Name(), ".link") {
-			continue
+
+	// NOTE: We assume that the base directory used to hold the links are unique
+	// (e.g., identified by the digest of the first link metadata file corresponding to the first step).
+	// This is so that different versions of links corresponding to the same root layout can be safely isolated from each other.
+	// TODO: If they are not unique, should we raise an error later when adding them to the TUF targets metadata?
+	digest := filepath.Base(dir)
+	links := make(Links)
+	for _, fileinfo := range fileinfos {
+		filename := fileinfo.Name()
+		if !strings.Contains(filename, ".link") {
+			return nil, fmt.Errorf("%s does not have a .link suffix", filename)
```

Comment:
Do we have to assume that every file in the directory must be a link? I could see someone wanting to have another file, such as a readme that may provide additional documentation or info about the files in the directory. Does it make sense to instead skip files that aren't links?